### PR TITLE
[Tests] Speed up some fast pipeline tests

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Currently, the fast pipeline tests take about 20 minutes to fully execute. https://github.com/huggingface/diffusers/actions/runs/8430950874/artifacts/1358187212 gives an idea. 

Below are the tests that take more than 10 seconds to run:

```bash
slowest durations
20.46s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_inference_batch_single_identical
20.17s call     tests/pipelines/animatediff/test_animatediff.py::AnimateDiffPipelineFastTests::test_inference_batch_single_identical
19.71s call     tests/pipelines/animatediff/test_animatediff.py::AnimateDiffPipelineFastTests::test_vae_slicing
19.43s call     tests/pipelines/stable_diffusion_2/test_stable_diffusion_attend_and_excite.py::StableDiffusionAttendAndExcitePipelineFastTests::test_karras_schedulers_shape
14.49s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_dict_tuple_outputs_equivalent
14.37s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_ip_adapter_single
14.28s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_ip_adapter_multi
13.95s call     tests/pipelines/animatediff/test_animatediff.py::AnimateDiffPipelineFastTests::test_save_load_local
13.90s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_save_load_optional_components
13.55s call     tests/pipelines/animatediff/test_animatediff.py::AnimateDiffPipelineFastTests::test_dict_tuple_outputs_equivalent
13.50s call     tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_save_load_local
13.25s call     tests/pipelines/animatediff/test_animatediff.py::AnimateDiffPipelineFastTests::test_save_load_optional_components
10.60s call     tests/pipelines/audioldm2/test_audioldm2.py::AudioLDM2PipelineFastTests::test_audioldm2_num_waveforms_per_prompt
10.55s call     tests/pipelines/stable_diffusion_xl/test_stable_diffusion_xl_adapter.py::StableDiffusionXLAdapterPipelineFastTests::test_StableDiffusionMixin_component
```

IMO we should try to optimize them a little. This PR is a first attempt at that. I tried a lot but couldn't meaningfully optimize the other tests that fall under this category. LMK if you you folks have any suggestions.

### CODE REVIEW

**Comment by yiyixuxu** [comment]
thanks for starting this! @sayakpaul 

for ip_adapter tests, I wonder if instead of getting a baseline output by running the pipeline without the ip-adapter, https://github.com/huggingface/diffusers/blob/544710ef0fe83c3ea87744c21a550d46e088a3c4/tests/pipelines/test_pipelines_common.py#L251 I think maybe we can skip this and just compare it against a static slice 

same can be applied in other places

**Comment by sayakpaul** [comment]
Love the idea, @yiyixuxu. Let me work on this in this PR itself. Thank you!

**Comment by sayakpaul** [comment]
@yiyixuxu I made some changes to check against static slices for "test_ip_adapter_single". If the plan looks okay to you, I will do the same for "test_ip_adapter_multi". And then in a future PR, I will tackle the following

> same can be applied in other tests e.g. test_dict_tuple_outputs_equivalent, test_inference_batch_single_identical

I think this will be easier for the reviewers too. 

Have left some comments in line to explain some thoughts. 

LMK.

File: src/diffusers/utils/testing_utils.py [hunk 1]
```diff
@@ -105,10 +105,21 @@ def numpy_cosine_similarity_distance(a, b):
...
+    expected_tensor_name="expected_slice",
+):
+    if max_torch_print:
+        torch.set_printoptions(threshold=10_000)
+
     test_name = os.environ.get("PYTEST_CURRENT_TEST")
     if not torch.is_tensor(tensor):
         tensor = torch.from_numpy(tensor)
+    if limit_to_slices:
+        tensor = tensor[0, -3:, -3:, -1]
```
Comment (sayakpaul) [inline src/diffusers/utils/testing_utils.py:122]: Convenience options. They are not harmful.

File: tests/pipelines/animatediff/test_animatediff.py [hunk 2]
```diff
@@ -131,6 +131,42 @@ def test_motion_unet_loading(self):
     def test_attention_slicing_forward_pass(self):
         pass
 
+    def test_ip_adapter_single(self):
+        expected_pipe_slice = None
+        if torch_device == "cpu":
```
Comment (sayakpaul) [inline tests/pipelines/animatediff/test_animatediff.py:136]: We make use of `torch_device` which can change based on the environment. The slices below were obtained on Intel CPUs which is our default mode and used on the PR tests too. 

For accelerators, the tests will run much faster anyway (even on MPS). So, I think this is the way to go here. But I welcome any other ideas too.

**Comment by DN6** [comment]
The block output sizes in the models can also be reduced. A lot of them use 32 and 64. These can be reduce to 4 and 8

**Comment by sayakpaul** [comment]
> The block output sizes in the models can also be reduced. A lot of them use 32 and 64. These can be reduce to 4 and 8

Agree. Could you take care of them in a separate PR? @DN6

**Comment by sayakpaul** [comment]
From the CI:

> https://github.com/huggingface/diffusers/actions/runs/8447126617/job/23136985269?pr=7477#step:6:12714

It's passing on the CPU GCP VM I am testing this on. I can increase the tolerance to `9e-4`.

**Comment by yiyixuxu** [comment]
how come it takes 23 min now?

File: tests/pipelines/controlnet/test_controlnet_sdxl.py [hunk 3]
```diff
@@ -191,6 +191,15 @@ def get_dummy_inputs(self, device, seed=0):
...
         return self._test_attention_slicing_forward_pass(expected_max_diff=2e-3)
 
+    def test_ip_adapter_single(self, from_ssd1b=False, expected_pipe_slice=None):
+        if not from_ssd1b:
+            expected_pipe_slice = None
+            if torch_device == "cpu":
+                expected_pipe_slice = np.array(
+                    [0.7331, 0.5907, 0.5667, 0.6029, 0.5679, 0.5968, 0.4033, 0.4761, 0.5090]
+                )
+        return super().test_ip_adapter_single(expected_pipe_slice=expected_pipe_slice)
```
Comment (yiyixuxu) [inline tests/pipelines/controlnet/test_controlnet_sdxl.py:201]: should we pass device = "cpu" to test_ip_adapter_single when using slice? would it help precision?
I saw a test failing right now ......

**Review by yiyixuxu — COMMENTED** [review]
very nice!
I saw tests/pipelines/pia/test_pia.py::PIAPipelineFastTests::test_ip_adapter_single went from 14 seconds to 9.97

**Comment by yiyixuxu** [comment]
cc @DN6 for a final review 

we can open this task to the community too! 
> If the plan looks okay to you, I will do the same for "test_ip_adapter_multi". And then in a future PR, I will tackle the following

> same can be applied in other tests e.g. test_dict_tuple_outputs_equivalent, test_inference_batch_single_identical

**Review by DN6 — APPROVED** [review]
Cool with me. Additionally, if you drop the `cross_attention_dim` in the dummy components it should lead to a big speedup for all these tests.

File: tests/pipelines/controlnet/test_controlnet_sdxl.py (hunk 3 above)
Comment (sayakpaul) [reply tests/pipelines/controlnet/test_controlnet_sdxl.py:201]: See: https://github.com/huggingface/diffusers/pull/7477#issuecomment-2022084100.

`expected_slice` is always None when the `torch_device` is not CPU.

**Comment by sayakpaul** [comment]
> cc @DN6 for a final review
> 
> we can open this task to the community too!
> 
> > If the plan looks okay to you, I will do the same for "test_ip_adapter_multi". And then in a future PR, I will tackle the following
> 
> > same can be applied in other tests e.g. test_dict_tuple_outputs_equivalent, test_inference_batch_single_identical

@yiyixuxu I feel like this is kind of a good thing for us to completely tackle because it would greatly improve the CI experience not only just for the contributors but also for us, mantainers. So, I prefer to do this myself. 

> Cool with me. Additionally, if you drop the `cross_attention_dim` in the dummy components it should lead to a big speedup for all these tests.

@DN6 could you tackle this and https://github.com/huggingface/diffusers/pull/7477#issuecomment-2022055045 in a PR?

**Comment by sayakpaul** [comment]
Okay so, the current failing test is different from the one I saw yesterday. The tolerance isn't that much to up, I believe. LMK.

**Pull request merged**